### PR TITLE
Ecomm

### PR DESCRIPTION
**Pull request by karl-young**
This PR adds automated tests for Typescript React Shopping Site.
**Key Features:**
- Tests for specific stock levels for all sizes.
- Tests for adding multiple items to the cart and verifying the item count.
- Tests for removing items from the cart.
- Tests for checking the total price is correct.
- A test that verifies the checkout process with a dialog confirmation.

## Test Results
- Verify stock levels are displayed correctly for each size filter: PASS
- Verify adding items to the cart works: PASS
- Verify removing an item from the cart works: PASS
- Verify the cart total is correct after adding and removing items: PASS
- Verify the checkout process works correctly: PASS

### CODE REVIEW

File: Pages/ShoppingSite.ts
Comment (TheFreeRangeTester) [inline Pages/ShoppingSite.ts:1]: A general recommendation: Name your page object classes with "Page" at the end. Like: MainPage.ts for this one.

File: Pages/ShoppingSite.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      await this.clickAddToCart(i)
+      await this.clickCloseCheckout()
+    }
+  }
+
+  async clickOpenCheckout() {
+    await this.openCheckout.click()
+  }
+
+  async ClickRemoveItem(index: number) {
```
Comment (TheFreeRangeTester) [inline Pages/ShoppingSite.ts]: camelCase :)

File: tests/shoppingSite.spec.ts
Comment (TheFreeRangeTester) [inline tests/shoppingSite.spec.ts:1]: For the steps, try to put yourself in the shoes of a manager or someone who wants to read a high level description of what's being done. This means a higher level to be written on the steps, as more low level details such as "retrieve stock level" doesn't tell a good story. 

As a potential customer
When I filter for a size of my choice
I can see results matching the criteria I selected.

**Review by TheFreeRangeTester — COMMENTED** [review]
Great job Karl! Some tweaks here and there, more on the soft skills level of communication than the technical aspect of the code.

File: Pages/ShoppingSite.ts
Comment (karl-young) [reply Pages/ShoppingSite.ts:1]: I will get on that.

File: Pages/ShoppingSite.ts (hunk 1 above)
Comment (karl-young) [reply Pages/ShoppingSite.ts:86]: I've started using that for naming folders too

Comment (karl-young) [reply Pages/ShoppingSite.ts]: Fixed

File: tests/shoppingSite.spec.ts
Comment (karl-young) [reply tests/shoppingSite.spec.ts:1]: Ok i have hopefully changed it so it has a better story .

File: Pages/ShoppingSite.ts
Comment (karl-young) [reply Pages/ShoppingSite.ts:1]: Changed and that's a good point.

File: tests/shoppingSite.spec.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      XS: 1,
+      S: 2,
+      M: 1,
+      ML: 2,
+      L: 10,
+      XL: 10,
+      XXL: 4,
+    }
+    for (const [size, expectedStock] of Object.entries(expectedStockLevels)) {
+      await test.step(`Given the customer has navigated to the site`, async () => {})
```
Comment (TheFreeRangeTester) [inline tests/shoppingSite.spec.ts]: We can skip an empty step by stating in the next one that the customer is performing the action from the page he is supposed to be at.

File: tests/shoppingSite.spec.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      await test.step(`Then the stock level for the "${size}" size should be ${expectedStock}.`, async () => {
+        const actualStock = await shoppingSite.getStockLevel()
+        console.log(`The stock level of ${size} is ${actualStock}.`)
+        expect(actualStock).toBe(expectedStock)
+      })
+    }
+  })
+
+  test('Verify adding items to the cart works', async ({ page }) => {
+    await test.step('Given the customer has selected 6 items to add to the cart', async () => {
```
Comment (TheFreeRangeTester) [inline tests/shoppingSite.spec.ts:38]: We can skip an empty step by stating in the next one that the customer is performing the action from the page he is supposed to be at.

File: tests/shoppingSite.spec.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      await shoppingSite.clickOpenCheckout()
+    })
+
+    await test.step('Then the cart should display 6 items', async () => {
+      expect(await shoppingSite.countItemsInCart()).toBe(6)
+    })
+  })
+
+  test('Verify removing an item from the cart works', async ({ page }) => {
+    await test.step('Given the customer has added an item to the cart', async () => {
```
Comment (TheFreeRangeTester) [inline tests/shoppingSite.spec.ts:51]: We can skip an empty step by stating in the next one that the customer is performing the action from the page he is supposed to be at.

**Review by TheFreeRangeTester — APPROVED** [review]

**Pull request merged**